movement/movement.h

```diff
@@ -312,4 +313,85 @@ void movement_play_alarm_beeps(uint8_t rounds, BuzzerNote alarm_note);
+ * durations and schedule future tasks. The HPT is not affected by changes to 
+ * the RTC clock time. While enabled, it continues to run in standby mode and 
+ * may be used to wake the watch up from sleep.
+ * 
+ * The HPT is disabled by default to conserve power. Before using it to get a
+ * timestamp, a watch face must activate it using "movement_hpt_request". If
+ * not already running, this will enable and start the timer module. While the
+ * timer is enabled, the face may retrieve the current timestamp using
+ * "movement_hpt_get", or schedule a background event using
+ * "movement_hpt_schedule". When a face no longer needs to use the timestamp or
+ * scheduled event provided by the HPT it MUST call "movement_hpt_relenquish".
+ * If no other face has an outstanding request for the HPT, it will be disabled
+ * to conserve power.
+ * 
+ * Watch faces may not modify the value of the HPT counter in any way. The only
+ * guarantee to be made about the HPT timestamp is that between the time your
+ * face calls "movement_hpt_request" until the moment it calls
+ * "movement_hpt_relenquish", the value returned from "movement_hpt_get" will
+ * increment upwards at 1024hz. Outside of that window, the timestamp value may
+ * change unpredictably.
+ * 
+ * Faces may schedule an EVENT_HPT event to occur by calling 
+ * "movement_hpt_schedule" and passing in a timestamp for the event to occur.
+ * The face must call "movement_hpt_request" before scheduling the event, and
+ * must not call "movement_hpt_relenquish" until after the event has occurred.
+ * Note that when your face receives the EVENT_HPT event, it may be running in
+ * the background. In this case, you will need to use the "_face" variant of
+ * the HPT methods to specify that it is your face being called.
+*/
+
+/**
+ * Enables the HPT for the active face. This method must be called before using
+ * "movement_hpt_get" or "movement_hpt_schedule". The HPT will remain running
+ * in the background until it is released using "movement_hpt_relenquish"
+*/
+void movement_hpt_request(void);
+/**
+ * A variant of "movement_hpt_request" that can be used when your face is not
+ * running in the foreground.
+*/
+void movement_hpt_request_face(uint8_t face_idx);
+
+/**
+ * Disables the HPT if no other faces are using it. This method must be called
+ * when your face no longer needs the timestamp provided by 
+ * "movement_hpt_get" or has no scheduled background tasks, in order to save
+ * power.
+*/
+void movement_hpt_relenquish(void);
+/**
+ * A variant of "movement_hpt_relenquish" that can be used when your face is
+ * not running in the foreground.
+*/
+void movement_hpt_relenquish_face(uint8_t face_idx);
+
+/**
+ * Schedules a future EVENT_HPT event to occur on or after the given timestamp.
+ * The Movement framework will do its best to trigger the event as close to the
+ * timestamp as possible, but it may be delayed if multiple faces or events are
+ * scheduled to occur on the same timestamp.
```

Comment:
This implies timers are guaranteed to be handled, which means either real time code or a queue of pending timer events. Do you mean to guarantee this, and if so by which method?

I assume you meant real time constraints, as implied by your concerns in discussion #381:

> If interrupt handling takes too long, it's possible for the timer value to change in the background while we're handling scheduled events. If events are scheduled too close together or too close in the future, it might be possible that by the time we set the timestamp value for the new interrupt, the counter has already passed it and the interrupt won't trigger. Now, if the counter is running at 128hz and the cpu is running at 4mhz, that gives us about 31,000 clock cycles to finish up whatever we were doing and check the timer again. Plenty of time... unless a face calls _delay_ms() or has to wait to activate a peripheral. Not a deal-breaking problem, but something we need to be aware of.

However this comment was written before you updated your timer code to 1024 Hz. The deadlines will be much tighter.

Reply:
I think the Movement framework needs to guarantee these events, but I haven't figured out exactly how to do it. It might be messy.

Reply:
We could check the nearest event to see if its timestamp is in the past. If so, that would mean its deadline was missed. We move forward in the event queue, processing them all in sequence until we find one with a timestamp in the future.

An efficient, hopefully constant time "next event" function would make it feasible. We'd be able to guarantee it would finish before the next timer window.